### ISSUE TITLE
[2019-12-8] Self-Written BatchNorm accumulate RunningVar Faulty

### ISSUE THREAD

**Issue by A-suozhang** [A-suozhang/my-debug-record#2]
## Bug Report
* 自实现的BN函数在前向和反向都和nn.BatchNorm2d相同的情况下,running mean积累正常,而running var积累错误

![](https://github.com/A-suozhang/MyPicBed/raw/master/img/20191208121731.png)

**Comment by A-suozhang** [comment]
## Initial File

* ![](https://github.com/A-suozhang/MyPicBed/raw/master/img/20191208121829.png)
    * 其中 bn0　为完全自己实现的(包括backward)
    * 其中 bn1  为pytorch原生
    * 其中 bn2 为实现了前向,后向由autograd产生

**Comment by A-suozhang** [comment]
* 再次参考了孙博的实现发现了问题所在
* 还是由于Bessel矫正(对于由样本方差估计总体方差的时候,将分母上的N变为N-1使其大一些)
  * ![](https://github.com/A-suozhang/MyPicBed/raw/master/img/20191208122146.png)
  * 而直接计算总体方差的时候分母仍然是N(与物理意义符合)
* pytorch原生的tensor.var方法加入了Bellman矫正;
* BN中前向时候用的是没有经过Bellman矫正的方差,而running var积累的时候用到了经过Bellman矫正之后的反差,与物理意义符合,所以需要修改running var的积累,手动乘上一个 N/(N-1)

## Solution

* ![](https://github.com/A-suozhang/MyPicBed/raw/master/img/20191208122444.png)
